gen/AArch64Arch_gen.ml

```diff
@@ -100,6 +104,14 @@ module SIMD = struct
       else reg r 0::regs (r+1) in
     regs 0
 
+  let reduce vec =
+    match vec with
+    | [] -> 0
+    | _ -> let sum = List.map2 (+) in
+           let init = [0;0;0;0] in
+           let v = (List.fold_right sum vec init) in
+           List.fold_right (+) v 0
```

Comment:
Ahh, did not know about that. I hope I got it right

```  let reduce vec =
    match vec with
    | [] -> 0
    | _ -> Misc.fold_cross vec (fun v k -> List.fold_right (+) v k) 0
```

Reply:
a fold cross should do both of the folds, so you shouldn't need another fold normally. I think you actually need to use Misc.fold_cross_gen because misc.fold_cross appends a list but you want to do an addition.

i think the following should work, but I haven't tested it. If it doesn't work let me know and I am happy to figure it out :)

Misc.fold_cross_gen (+) 0 vec (+) 0

Reply:
Hmm, it seems that `Misc.fold_cross` works differently to my original proposal. 
Before `Misc.fold_cross`
```
 diyone7 -arch AArch64 -variant neon "Rfe PodRW Rfe DpDatadWNe2P"
AArch64 A
"Rfe PodRW RfePNe2 DpDatadWNe2P"
Generator=diyone7 (version 7.56+03)
Prefetch=0:x=F,0:y=W,1:y=F,1:x=W
Com=Rf Rf
Orig=Rfe PodRW RfePNe2 DpDatadWNe2P
{
int y[8];

0:X0=x; 0:X3=y;
1:X0=y; 1:X3=x;
}
 P0          | P1                      ;
 LDR W1,[X0] | LD1 {V0.4S, V1.4S},[X0] ;
 MOV W2,#1   | ADD V0.4S,V0.4S,V1.4S   ;
 STR W2,[X3] | ADDV S2,V0.4S           ;
             | FMOV W1,S2              ;
             | EOR W2,W1,W1            ;
             | ADD W2,W2,#1            ;
             | STR W2,[X3]             ;
exists (0:X1=1 /\ 1:X1=1)
 ```
 Note, `1:X1=1` which looks resalable since we only store value '1' 
 After change to `Misc.fold_cross`
 ```
  diyone7 -arch AArch64 -variant neon "Rfe PodRW Rfe DpDatadWNe2P"
AArch64 A
"Rfe PodRW RfePNe2 DpDatadWNe2P"
Generator=diyone7 (version 7.56+03)
Prefetch=0:x=F,0:y=W,1:y=F,1:x=W
Com=Rf Rf
Orig=Rfe PodRW RfePNe2 DpDatadWNe2P
{
int y[8];

0:X0=x; 0:X3=y;
1:X0=y; 1:X3=x;
}
 P0          | P1                      ;
 LDR W1,[X0] | LD1 {V0.4S, V1.4S},[X0] ;
 MOV W2,#1   | ADD V0.4S,V0.4S,V1.4S   ;
 STR W2,[X3] | ADDV S2,V0.4S           ;
             | FMOV W1,S2              ;
             | EOR W2,W1,W1            ;
             | ADD W2,W2,#1            ;
             | STR W2,[X3]             ;
exists (0:X1=1 /\ 1:X1=4)
```
Note, now `1:X1=4` which seems impossible

Reply:
I think it can be as simple as
```
let reduce vec = List.fold_right (+) (List.flatten vec) 0
```

Reply:
Perfect! That looks good to me